File: shell/src/bridge/facade.test.ts
// The facade call path, against a fake module. What is under test is the marshalling contract —
// name resolution, heap handling, failure decoding — not the core's behaviour, which has its own
// suite natively.
import { describe, expect, it, vi } from 'vitest';

import { createFacadeCall, decodeStatus, type FacadeModule } from './facade';
import { Reader, Writer } from './wire';

/** A module that echoes a scripted response and records what it was handed. */
function fakeModule(options: {
  methods?: string[];
  respond?: (id: number, args: Uint8Array<ArrayBufferLike>) => Uint8Array<ArrayBufferLike>;
  growHeapOnCall?: boolean;
} = {}) {
  const methods = options.methods ?? ['ProjectManager.create', 'ProjectManager.list'];
  let heap = new Uint8Array(4096);
  const nameOffsets = new Map<number, string>();
  let next = 1024;
  let lastResult: Uint8Array<ArrayBufferLike> = new Uint8Array(0);
  const calls: Array<{ id: number; args: Uint8Array<ArrayBufferLike> }> = [];
  const freed: number[] = [];

  const module: FacadeModule & { calls: typeof calls; freed: number[] } = {
    get HEAPU8() {
      return heap;
    },
    UTF8ToString: (pointer: number) => nameOffsets.get(pointer) ?? '',
    _sph_facade_method_count: () => methods.length,
    _sph_facade_method_name: (id: number) => {
      const pointer = 100 + id;
      nameOffsets.set(pointer, methods[id] ?? '');
      return pointer;
    },
    _sph_facade_call: (id: number, argsPointer: number, length: number) => {
      const args = heap.slice(argsPointer, argsPointer + length);
      calls.push({ id, args });
      // Emscripten swaps HEAPU8 for a new view when memory grows; a caller holding the old one
      // is reading a detached buffer.
      if (options.growHeapOnCall) heap = new Uint8Array(heap.length * 2);
      lastResult = options.respond?.(id, args) ?? new Uint8Array(0);
      heap.set(lastResult, 2048);
      return lastResult.length;
    },
    _sph_facade_result: () => 2048,
    _malloc: (bytes: number) => {
      const pointer = next;
      next += bytes;
      return pointer;
    },
    _free: (pointer: number) => {
      freed.push(pointer);
    },
    calls,
    freed,
  };
  return module;
}

function okStatus(): Uint8Array {
  const w = new Writer();
  w.i32(0);          // StatusCode.Ok
  w.string('');
  w.string('');
  return w.finish();
}

describe('createFacadeCall', () => {
  it('resolves method names to ids once, at construction', () => {
    const module = fakeModule();
    const spy = vi.spyOn(module, '_sph_facade_method_name');
    const call = createFacadeCall(module);
    const before = spy.mock.calls.length;
    void call('ProjectManager.list', new Uint8Array(0));
    expect(spy.mock.calls.length).toBe(before);
  });

  it('calls the id the core published for that name', async () => {
    const module = fakeModule({ respond: () => okStatus() });
    const call = createFacadeCall(module);
    await call('ProjectManager.list', new Uint8Array(0));
    expect(module.calls[0]?.id).toBe(1);
  });

  it('passes the encoded arguments through unchanged', async () => {
    const module = fakeModule({ respond: () => okStatus() });
    const call = createFacadeCall(module);
    await call('ProjectManager.create', new Uint8Array([1, 2, 3, 4]));
    expect(Array.from(module.calls[0]!.args)).toEqual([1, 2, 3, 4]);
  });

  it('rejects a method the core does not have', async () => {
    // A client bundle can be older or newer than the core it loaded. Failing at the call site
    // names the method; a silent no-op would surface as an empty result much later.
    const call = createFacadeCall(fakeModule());
    await expect(call('ProjectManager.nope', new Uint8Array(0)))
      .rejects.toThrow(/ProjectManager\.nope/);
  });

  it('frees the argument buffer it allocated', async () => {
    const module = fakeModule({ respond: () => okStatus() });
    const call = createFacadeCall(module);
    await call('ProjectManager.create', new Uint8Array([1, 2, 3]));
    expect(module.freed.length).toBe(1);
  });

  it('frees the argument buffer even when the call fails', async () => {
    const module = fakeModule({
      respond: () => { throw new Error('core trapped'); },
    });
    const call = createFacadeCall(module);
    await expect(call('ProjectManager.create', new Uint8Array([1]))).rejects.toThrow();
    expect(module.freed.length).toBe(1);
  });

  it('copies the result out of the heap rather than viewing into it', async () => {
    // The core reuses one result buffer, and Emscripten replaces HEAPU8 wholesale when memory
    // grows. A view into either would read someone else's bytes on the next call.
    const module = fakeModule({ growHeapOnCall: true, respond: () => okStatus() });
    const call = createFacadeCall(module);
    const first = await call('ProjectManager.list', new Uint8Array(0));
    await call('ProjectManager.list', new Uint8Array(0));
    expect(first.length).toBe(okStatus().length);
    expect(Array.from(first)).toEqual(Array.from(okStatus()));
  });

  it('handles an empty argument list without allocating nothing', async () => {
    const module = fakeModule({ respond: () => okStatus() });
    const call = createFacadeCall(module);
    const result = await call('ProjectManager.list', new Uint8Array(0));
    expect(result.length).toBeGreaterThan(0);
  });
});

describe('an allocation the core cannot satisfy', () => {
  it('reports a failed malloc instead of writing to heap offset zero', async () => {
    // With memory growth on, _malloc returns 0 when it cannot satisfy a request — which on a
    // phone mid-capture, with a sphere of frames already pinned, is a real outcome rather than a
    // theoretical one. Treating that 0 as an address writes the arguments over the start of the
    // heap and then calls into C++ with a null pointer: memory corruption surfacing as whatever
    // fails next, instead of as the allocation failure it is.
    const module = fakeModule();
    module._malloc = () => 0;

    const call = createFacadeCall(module);
    await expect(call('ProjectManager.list', new Uint8Array([1, 2, 3])))
      .rejects.toThrow(/allocate/i);
    // Nothing crossed the boundary, and nothing was freed that was never allocated.
    expect(module.calls).toHaveLength(0);
    expect(module.freed).toHaveLength(0);
  });

  it('still frees the buffer when the call itself fails', async () => {
    const module = fakeModule();
    module._sph_facade_call = () => { throw new Error('trapped'); };

    const call = createFacadeCall(module);
    await expect(call('ProjectManager.list', new Uint8Array())).rejects.toThrow('trapped');
    expect(module.freed).toHaveLength(1);
  });
});

describe('decodeStatus', () => {
  it('reads a success status', () => {
    const status = decodeStatus(new Reader(okStatus()));
    expect(status.code).toBe('Ok');
  });

  it('reads a failure with its component and detail', () => {
    const w = new Writer();
    w.i32(1);                        // InvalidArgument
    w.string('ProjectManager');
    w.string('a project needs a title');
    const status = decodeStatus(new Reader(w.finish()));
    expect(status.code).toBe('InvalidArgument');
    expect(status.component).toBe('ProjectManager');
    expect(status.detail).toBe('a project needs a title');
  });

  it('refuses a truncated status rather than reading it as success', () => {
    // Every read past the end of a failed Reader returns zero, and zero is the index of Ok. A
    // core that trapped mid-call, or a result buffer that got clipped, would otherwise arrive as
    // a successful call with default-valued everything — the worst possible way to fail.
    const w = new Writer();
    w.i32(0);
    // no component, no detail: the payload stops here
    const status = decodeStatus(new Reader(w.finish()));
    expect(status.code).toBe('Internal');
    expect(status.detail).toMatch(/malformed/i);
  });

  it('refuses an empty response', () => {
    const status = decodeStatus(new Reader(new Uint8Array()));
    expect(status.code).toBe('Internal');
    expect(status.detail).toMatch(/malformed/i);
  });

  it('reads its code names from the generated table, not a copy of it', async () => {
    // A second hand-maintained list of the enum is the drift the generated boundary exists to
    // stop: adding a status in the C++ header would type-check here and silently shift the
    // meaning of every code after it.
    const codec = await import('./codec.generated');
    const w = new Writer();
    w.i32(codec.StatusCodeValues.indexOf('FrameStoreExhausted'));
    w.string('FrameStore');
    w.string('out of room');
    expect(decodeStatus(new Reader(w.finish())).code).toBe('FrameStoreExhausted');
  });

  it('does not invent a code for an index it does not know', () => {
    // A core newer than this bundle can return a status code this build has never heard of.
    const w = new Writer();
    w.i32(9999);
    w.string('X');
    w.string('from the future');
    const status = decodeStatus(new Reader(w.finish()));
    expect(status.code).toBe('Internal');
    expect(status.detail).toContain('from the future');
  });
});

File: bridge/test/facade_test.cpp
// The generated dispatch, exercised natively.
//
// The boundary is where a mistake is most expensive to find, and finding it in a browser is the
// slowest way to find anything. Everything here runs against the same generated switch the WASM
// build compiles.
#include <gtest/gtest.h>

#include <limits>
#include <set>
#include <string>
#include <vector>

#include "facade.h"
#include "sphanorama/codec.h"

namespace sphanorama {
namespace {

int32_t MethodId(const std::string& wireName) {
  for (int32_t id = 0; id < sph_facade_method_count(); ++id) {
    const char* name = sph_facade_method_name(id);
    if (name != nullptr && wireName == name) return id;
  }
  return -1;
}

struct Response {
  Status status;
  std::vector<uint8_t> bytes;
  wire::Reader reader() const { return wire::Reader(bytes.data(), bytes.size()); }
};

Response Call(const std::string& wireName, const std::vector<uint8_t>& args = {}) {
  const int32_t id = MethodId(wireName);
  EXPECT_GE(id, 0) << "no such method: " << wireName;
  const int32_t length = sph_facade_call(id, args.data(), static_cast<int32_t>(args.size()));
  Response response;
  response.bytes.assign(sph_facade_result(), sph_facade_result() + length);
  return response;
}

Status ReadStatus(wire::Reader& in) {
  Status status;
  status.code = static_cast<StatusCode>(in.GetI32());
  in.GetString();                     // component, not asserted on here
  status.detail = in.GetString();
  return status;
}

TEST(Facade, PublishesEveryMethodByName) {
  // A client resolves names at startup rather than hard-coding ids, which shift the day a method
  // is inserted above them.
  ASSERT_GT(sph_facade_method_count(), 0);
  std::set<std::string> names;
  for (int32_t id = 0; id < sph_facade_method_count(); ++id) {
    const char* name = sph_facade_method_name(id);
    ASSERT_NE(name, nullptr) << "method " << id;
    names.insert(name);
  }
  EXPECT_EQ(static_cast<int32_t>(names.size()), sph_facade_method_count());
  EXPECT_TRUE(names.count("ProjectManager.create"));
  EXPECT_TRUE(names.count("CaptureSessionManager.onMotion"));
}

TEST(Facade, OnlyManagersAreDispatched) {
  // Resource accesses are mirrored into TypeScript because the browser implements them; the call
  // goes the other way, so dispatching them here would mean calling a runtime that has no such
  // thing.
  for (int32_t id = 0; id < sph_facade_method_count(); ++id) {
    const std::string name = sph_facade_method_name(id);
    EXPECT_NE(name.find("Manager."), std::string::npos) << name;
  }
}

TEST(Facade, AnOutOfRangeNameIsNull) {
  EXPECT_EQ(sph_facade_method_name(-1), nullptr);
  EXPECT_EQ(sph_facade_method_name(sph_facade_method_count()), nullptr);
}

TEST(Facade, AnUnknownMethodIdIsReportedRatherThanCrashing) {
  // A cached client bundle can be older than the core it loaded: a version mismatch is something
  // to report, not a trap.
  const int32_t length = sph_facade_call(9999, nullptr, 0);
  ASSERT_GT(length, 0);
  std::vector<uint8_t> bytes(sph_facade_result(), sph_facade_result() + length);
  wire::Reader in(bytes.data(), bytes.size());
  EXPECT_EQ(ReadStatus(in).code, StatusCode::NotFound);
}

TEST(Facade, CreatesAProjectAndReadsItBack) {
  wire::Writer args;
  args.PutString("kitchen");
  Response created = Call("ProjectManager.create", args.bytes());

  wire::Reader in = created.reader();
  ASSERT_EQ(ReadStatus(in).code, StatusCode::Ok);
  const auto id = static_cast<uint64_t>(in.GetF64());
  EXPECT_GT(id, 0u);

  Response listed = Call("ProjectManager.list");
  wire::Reader listing = listed.reader();
  ASSERT_EQ(ReadStatus(listing).code, StatusCode::Ok);
  const size_t count = listing.GetCount(1);
  ASSERT_GE(count, 1u);

  bool found = false;
  for (size_t i = 0; i < count; ++i) {
    ProjectSummary summary;
    ASSERT_TRUE(codec::Decode(listing, summary));
    if (summary.id.value == id) {
      EXPECT_EQ(summary.title, "kitchen");
      found = true;
    }
  }
  EXPECT_TRUE(found);
}

TEST(Facade, ASelectionCrossesTheBoundaryAndComesBackTheSameNumber) {
  // The new call in ADR 0040, over the generated dispatch rather than over a manager held
  // directly. Until now the only thing exercising it was one end-to-end test in a browser, which
  // is the slowest place to find a boundary mistake and the one that says least about which end
  // made it.
  //
  // Each `Response` is held in a named local because `reader()` points into its buffer, which a
  // temporary would free at the end of the statement.
  wire::Writer named;
  named.PutString("kitchen");
  Response opened = Call("ProjectManager.create", named.bytes());
  wire::Reader created = opened.reader();
  ASSERT_EQ(ReadStatus(created).code, StatusCode::Ok);
  const double project = created.GetF64();

  // A cell nobody has chosen for is a *success* answering zero, not a failure — the sentinel this
  // whole decision turns on, and this is the only test that watches it survive the wire.
  wire::Writer cell;
  cell.PutF64(project);
  cell.PutF64(3.0);
  Response never = Call("ProjectManager.getSelection", cell.bytes());
  wire::Reader absent = never.reader();
  ASSERT_EQ(ReadStatus(absent).code, StatusCode::Ok);
  EXPECT_EQ(absent.GetF64(), 0.0);
  // The reader is asked whether it read anything, which on this leg is the whole assertion.
  // `StatusCode::Ok` is zero and a failed `wire::Reader` answers zero to everything, so both lines
  // above are also satisfied by a response of no bytes at all — on the one leg where zero is the
  // value under test rather than an incidental one. `wire.h` states this protocol: check `ok()`
  // once at the end rather than after every read.
  EXPECT_TRUE(absent.ok()) << "a zero read out of a failed reader is not an answer of zero";

  wire::Writer pick;
  pick.PutF64(project);
  pick.PutF64(3.0);
  pick.PutF64(7.0);
  Response wrote = Call("ProjectManager.setSelection", pick.bytes());
  wire::Reader recorded = wrote.reader();
  ASSERT_EQ(ReadStatus(recorded).code, StatusCode::Ok);

  Response asked = Call("ProjectManager.getSelection", cell.bytes());
  wire::Reader read = asked.reader();
  ASSERT_EQ(ReadStatus(read).code, StatusCode::Ok);
  EXPECT_EQ(read.GetF64(), 7.0);
  EXPECT_TRUE(read.ok());

  // And a cell the writer refuses is refused rather than answered, so the two ends cannot be made
  // to agree on a value neither of them considers an identity.
  wire::Writer unset;
  unset.PutF64(project);
  unset.PutF64(0.0);
  Response denied = Call("ProjectManager.getSelection", unset.bytes());
  wire::Reader refused = denied.reader();
  EXPECT_EQ(ReadStatus(refused).code, StatusCode::InvalidArgument);
}

TEST(Facade, ADomainFailureCrossesAsAStatusNotAsATrap) {
  wire::Writer args;
  args.PutString("");                 // an untitled project is refused by the manager
  Response response = Call("ProjectManager.create", args.bytes());
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument);
}

TEST(Facade, MalformedArgumentsAreRejectedRatherThanDecodedAsGarbage) {
  // What a truncated postMessage looks like from the receiving side.
  const std::vector<uint8_t> truncated{0x40, 0x00};
  Response response = Call("ProjectManager.create", truncated);
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument);
}

TEST(Facade, TheDetailExplainsWhyRatherThanRepeatingTheCode) {
  wire::Writer args;
  args.PutString("");
  Response response = Call("ProjectManager.create", args.bytes());
  wire::Reader in = response.reader();
  EXPECT_FALSE(ReadStatus(in).detail.empty());
}

TEST(Facade, StartingACaptureSessionFailsHonestlyWithoutACamera) {
  // The camera lives in JavaScript and nothing on this side opened one. The session refuses with
  // a reason rather than producing empty frames, which is the whole argument for null over stub.
  wire::Writer titled;
  titled.PutString("a project to capture into");
  Response created = Call("ProjectManager.create", titled.bytes());
  wire::Reader out = created.reader();
  ASSERT_EQ(ReadStatus(out).code, StatusCode::Ok);
  const double project = out.GetF64();

  wire::Writer args;
  args.PutF64(project);
  CapturePlanSpec spec;
  spec.acceptanceConeDeg = 4.0;
  codec::Encode(args, spec);

  Response response = Call("CaptureSessionManager.begin", args.bytes());
  wire::Reader in = response.reader();
  const Status status = ReadStatus(in);
  EXPECT_EQ(status.code, StatusCode::CameraUnavailable);
}

TEST(Facade, ATruncatedStructArgumentIsReportedAsAStatusNotAnEmptyBuffer) {
  // The client may be a stale bundle, a corrupted transfer, or something hostile. Whatever the
  // cause, the ABI promises a status — and a decode that left the switch case early handed back
  // an empty buffer instead, which the client can only report as "malformed response".
  wire::Writer args;
  args.PutF64(1.0);          // node id
  args.PutF64(5.0);          // frameCount, then the BurstSpec simply stops
  Response response = Call("CaptureSessionManager.armBurst", args.bytes());
  ASSERT_FALSE(response.bytes.empty()) << "the facade returned nothing at all";
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument);
}

TEST(Facade, APreviewRequestReachesTheManagerRatherThanStoppingAtTheArguments) {
  // The one call that answers with pixels (ADR 0038), and the first with an `int32_t` parameter:
  // the generator used to decode every number into a `double`, which the core will not narrow.
  // Three arguments in the order the contract declares them, and a status back from the manager
  // — `FailedPrecondition`, because nothing here has begun a session — rather than the facade's
  // own `InvalidArgument`, which is what a parameter list read in the wrong order produces.
  wire::Writer args;
  args.PutF64(1.0);          // node
  args.PutF64(1.0);          // candidate
  args.PutF64(128.0);        // maxEdge
  Response response = Call("CaptureSessionManager.candidatePreview", args.bytes());
  ASSERT_FALSE(response.bytes.empty()) << "the facade returned nothing at all";
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::FailedPrecondition);
}

TEST(Facade, ATruncatedPreviewRequestIsReportedAsAStatus) {
  wire::Writer args;
  args.PutF64(1.0);          // a node, and then the request simply stops
  Response response = Call("CaptureSessionManager.candidatePreview", args.bytes());
  ASSERT_FALSE(response.bytes.empty());
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument);
}

TEST(Facade, AnIntegerArgumentThatIsNotOneIsRefusedRatherThanCast) {
  // Every number crosses as a double, because JavaScript has no other kind, and an `int32_t`
  // parameter is therefore a double that has to become an integer. `static_cast` of a NaN or of
  // anything outside the target's range is undefined — the same hazard `GetId` exists for, one
  // door along. So the reader refuses the message rather than converting it, and the caller gets
  // the facade's own `InvalidArgument` instead of whatever the cast happened to produce.
  for (const double raw : {std::numeric_limits<double>::quiet_NaN(),
                           std::numeric_limits<double>::infinity(),
                           1e300, -1e300, 2147483648.0, -2147483649.0, 12.5}) {
    wire::Writer args;
    args.PutF64(1.0);          // node
    args.PutF64(1.0);          // candidate
    args.PutF64(raw);          // maxEdge, which is an int32_t on the other side
    Response response = Call("CaptureSessionManager.candidatePreview", args.bytes());
    ASSERT_FALSE(response.bytes.empty()) << "the facade returned nothing at all for " << raw;
    wire::Reader in = response.reader();
    EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument) << "for " << raw;
  }
}

TEST(Facade, AnIntegerArgumentAtTheEdgeOfItsRangeStillCrosses) {
  // The bounds are inclusive, and asserting that matters: a check written with the wrong
  // comparison would reject the largest valid value and nothing above would notice.
  for (const double raw : {2147483647.0, -2147483648.0}) {
    wire::Writer args;
    args.PutF64(1.0);
    args.PutF64(1.0);
    args.PutF64(raw);
    Response response = Call("CaptureSessionManager.candidatePreview", args.bytes());
    ASSERT_FALSE(response.bytes.empty());
    wire::Reader in = response.reader();
    // Refused by the manager for having no session, which is one layer past the reader — the
    // point is that it got there rather than being rejected as malformed.
    EXPECT_EQ(ReadStatus(in).code, StatusCode::FailedPrecondition) << "for " << raw;
  }
}

TEST(Facade, AListCountIsRefusedWhenThePayloadCouldNotPossiblyHoldThatMany) {
  // The count is what the vector is sized from, so bounding it at one byte per element let a
  // small message ask for a very large allocation: 100 kB of filler claiming 100000 ImuSamples
  // is over 11 MB of vector on a phone, from a payload that cannot contain a single valid one.
  constexpr int32_t kClaimed = 100000;
  wire::Writer args;
  args.PutCount(kClaimed);
  std::vector<uint8_t> payload = args.bytes();
  payload.resize(payload.size() + kClaimed, 0);   // one byte per element promised

  Response response = Call("CaptureSessionManager.onMotion", payload);
  ASSERT_FALSE(response.bytes.empty());
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument);
}

TEST(Facade, ATruncatedListArgumentIsReportedAsAStatus) {
  wire::Writer args;
  args.PutCount(4);          // four samples promised, none supplied
  Response response = Call("CaptureSessionManager.onMotion", args.bytes());
  ASSERT_FALSE(response.bytes.empty());
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument);
}

TEST(Facade, AnIdThatIsNotAWholeNumberIsRefused) {
  // Ids cross as doubles because JavaScript has no other number type, so NaN, a negative and
  // 1e300 are all things a client can send. Converting any of them to uint64_t is undefined
  // behaviour, and it happens before a manager ever sees the id.
  for (const double bad : {std::numeric_limits<double>::quiet_NaN(),
                           -1.0, 0.5, 1e300}) {
    wire::Writer args;
    args.PutF64(bad);
    Response response = Call("ProjectManager.delete", args.bytes());
    ASSERT_FALSE(response.bytes.empty());
    wire::Reader in = response.reader();
    EXPECT_EQ(ReadStatus(in).code, StatusCode::InvalidArgument) << "for " << bad;
  }
}

TEST(Facade, ACaptureSessionForAProjectThatDoesNotExistIsRefused) {
  // Checked before the camera, so this is the answer even though no camera is open either.
  wire::Writer args;
  args.PutF64(4040.0);
  CapturePlanSpec spec;
  spec.acceptanceConeDeg = 4.0;
  codec::Encode(args, spec);

  Response response = Call("CaptureSessionManager.begin", args.bytes());
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::NotFound);
}

TEST(Facade, ABuildCannotBeStartedYetAndSaysSo) {
  wire::Writer args;
  args.PutF64(1.0);
  BuildSpec spec;
  codec::Encode(args, spec);
  Response response = Call("PanoramaBuildManager.start", args.bytes());
  wire::Reader in = response.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::Unsupported);
}

TEST(Facade, TheResultBufferSurvivesUntilTheNextCall) {
  wire::Writer args;
  args.PutString("first");
  Response first = Call("ProjectManager.create", args.bytes());
  ASSERT_FALSE(first.bytes.empty());

  Call("ProjectManager.list");
  // `first` copied its bytes, so it is still readable — which is the contract the TypeScript side
  // relies on when it copies out of HEAPU8 before making another call.
  wire::Reader in = first.reader();
  EXPECT_EQ(ReadStatus(in).code, StatusCode::Ok);
}

}  // namespace
}  // namespace sphanorama
